streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java

```diff
@@ -1103,32 +1156,67 @@ public void process(final ProcessorSupplier<? super K, ? super V> processorSuppl
 
-            final String joinThisName = rightOuter ?
-                    renamed.suffixWithOrElseGet("-outer-this-join", builder, OUTERTHIS_NAME)
-                    : renamed.suffixWithOrElseGet("-this-join", builder, JOINTHIS_NAME);
-            final String joinOtherName = leftOuter ?
-                    renamed.suffixWithOrElseGet("-outer-other-join", builder, OUTEROTHER_NAME)
-                    : renamed.suffixWithOrElseGet("-other-join", builder, JOINOTHER_NAME);
+            final String joinThisGeneratedName = rightOuter ? builder.newProcessorName(OUTERTHIS_NAME) : builder.newProcessorName(JOINTHIS_NAME);
+            final String joinOtherGeneratedName = leftOuter ? builder.newProcessorName(OUTEROTHER_NAME) : builder.newProcessorName(JOINOTHER_NAME);
+
+            final String joinThisName = renamed.suffixWithOrElseGet(joinThisSuffix, joinThisGeneratedName);
+            final String joinOtherName = renamed.suffixWithOrElseGet(joinOtherSuffix, joinOtherGeneratedName);
+
             final String joinMergeName = renamed.suffixWithOrElseGet(
                     "-merge", builder, MERGE_NAME);
+
             final StreamsGraphNode thisStreamsGraphNode = ((AbstractStream) lhs).streamsGraphNode;
             final StreamsGraphNode otherStreamsGraphNode = ((AbstractStream) other).streamsGraphNode;
 
 
-            final StoreBuilder<WindowStore<K1, V1>> thisWindowStore =
-                joinWindowStoreBuilder(joinThisName, windows, joined.keySerde(), joined.valueSerde());
-            final StoreBuilder<WindowStore<K1, V2>> otherWindowStore =
-                joinWindowStoreBuilder(joinOtherName, windows, joined.keySerde(), joined.otherValueSerde());
+            final StoreBuilder<WindowStore<K1, V1>> thisWindowStore;
+            final StoreBuilder<WindowStore<K1, V2>> otherWindowStore;
+            final String userProvidedBaseStoreName = materializedInternal == null ? null : materializedInternal.storeName();
+
+            final String thisJoinStoreName =  userProvidedBaseStoreName == null ? joinThisGeneratedName : userProvidedBaseStoreName + joinThisSuffix;
+            final String otherJoinStoreName = userProvidedBaseStoreName == null ? joinOtherGeneratedName : userProvidedBaseStoreName + joinOtherSuffix;
+
+            if (materializedInternal == null || materializedInternal.storeSupplier() == null) {
+                thisWindowStore = joinWindowStoreBuilder(thisJoinStoreName, windows, joined.keySerde(), joined.valueSerde());
+                otherWindowStore = joinWindowStoreBuilder(otherJoinStoreName, windows, joined.keySerde(), joined.otherValueSerde());
+            } else {
+                final WindowBytesStoreSupplier windowBytesStoreSupplier = (WindowBytesStoreSupplier) materializedInternal.storeSupplier();
+
+                final InMemoryWindowBytesStoreSupplier joinThisInMemorySupplier = new InMemoryWindowBytesStoreSupplier(windowBytesStoreSupplier.name() + joinThisSuffix,
```

Comment:
Just realized that I'm making an assumption here that a `StoreSupplier` here is meant to be used as an `InMemoryWindowBytesStoreSupplier`, but is this correct?  I'm thinking I need to refactor this to accept any `StoreSupplier`, but we need to create the 2nd instance to make sure the store names are unique.

Reply:
We should not assume that it's an in-memory store. We will only need a "WrapperSupplier" that wraps the original `WindowBytesStoreSupplier` but returns a modified name only -- all other method calls should be forwarded to the original/wrapper supplier).

We also need to add the corresponding runtime check though:
 - `windowBytesStoreSupplier.windowSize() == windows.size()`
 - `windowBytesStoreSupplier.retentionPeriod() >= windows.size()` (or `>= size + grace` ?)
 - `retainDuplicates() == false`
 - do we also need to check `segments()` and/or `segmentIntervalMs()` ?
 - check that `keySerde` and `valueSerde` from `Materialized` are both `null`

We should add corresponding unit tests (positive and negative) for each check.